File: verification/canonicalize.mjs
#!/usr/bin/env node
/**
 * CITIZEN_ROOT canonicalization.
 *
 * Rules (from CITIZEN_ROOT_INDEX verification_rules):
 *   canonicalization: UTF-8, LF, sorted object keys, no trailing whitespace
 *   hash_algorithm:   SHA-256
 *   leaf_construction: SHA-256(canonical_bytes)
 *   rewrite_forbidden: true
 *   authority: false
 *
 * Deterministic. No randomness. No authority. Fail closed.
 */

import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import { pathToFileURL } from "node:url";

function sha256Hex(data) {
  return createHash("sha256").update(data).digest("hex");
}

function stripTrailingWsLines(text) {
  const lines = text.split("\n");
  const stripped = lines.map((line) => line.replace(/[ \t\r\f\v]+$/g, ""));
  let body = stripped.join("\n");
  if (body && !body.endsWith("\n")) {
    body += "\n";
  }
  if (body.endsWith("\n\n")) {
    body = body.replace(/\n+$/, "\n");
  }
  return body;
}

export function canonicalizeText(raw) {
  let buf = Buffer.isBuffer(raw) ? raw : Buffer.from(raw, "utf8");
  if (buf.length >= 3 && buf[0] === 0xef && buf[1] === 0xbb && buf[2] === 0xbf) {
    buf = buf.subarray(3);
  }
  let text = buf.toString("utf8");
  if (text.charCodeAt(0) === 0xfeff) {
    text = text.slice(1);
  }
  text = text.replace(/\r\n/g, "\n").replace(/\r/g, "\n");
  text = stripTrailingWsLines(text);
  return Buffer.from(text, "utf8");
}

function sortKeysRecursive(obj) {
  if (obj !== null && typeof obj === "object" && !Array.isArray(obj)) {
    const out = {};
    for (const k of Object.keys(obj).sort()) {
      out[k] = sortKeysRecursive(obj[k]);
    }
    return out;
  }
  if (Array.isArray(obj)) {
    return obj.map(sortKeysRecursive);
  }
  return obj;
}

export function canonicalizeJson(raw) {
  let data;
  if (typeof raw === "object" && raw !== null && !Buffer.isBuffer(raw)) {
    data = raw;
  } else {
    let buf = Buffer.isBuffer(raw) ? raw : Buffer.from(String(raw), "utf8");
    if (buf.length >= 3 && buf[0] === 0xef && buf[1] === 0xbb && buf[2] === 0xbf) {
      buf = buf.subarray(3);
    }
    let text = buf.toString("utf8");
    if (text.charCodeAt(0) === 0xfeff) {
      text = text.slice(1);
    }
    text = text.replace(/\r\n/g, "\n").replace(/\r/g, "\n");
    data = JSON.parse(text);
  }
  const ordered = sortKeysRecursive(data);
  let serialized = JSON.stringify(ordered);
  serialized = serialized.replace(/[ \t\r\n]+$/g, "") + "\n";
  return Buffer.from(serialized, "utf8");
}

function detectJson(path, raw) {
  if (path && path.toLowerCase().endsWith(".json")) return true;
  const sample = raw.toString("utf8").trimStart().slice(0, 1);
  return sample === "{" || sample === "[";
}

export function canonicalize(raw, path = null, forceJson = null) {
  const buf = Buffer.isBuffer(raw) ? raw : Buffer.from(raw);
  const useJson = forceJson !== null ? forceJson : detectJson(path, buf);
  if (useJson) {
    if (forceJson === true) {
      // Explicit JSON: fail closed, never fall back
      return canonicalizeJson(buf);
    }
    try {
      return canonicalizeJson(buf);
    } catch {
      // Auto-detect only: treat as text
      return canonicalizeText(buf);
    }
  }
  return canonicalizeText(buf);
}

export function hashCanonical(raw, path = null, forceJson = null) {
  const canonical = canonicalize(raw, path, forceJson);
  return {
    path,
    sha256: sha256Hex(canonical),
    size_bytes: canonical.length,
    canonicalization: "UTF-8, LF, sorted object keys, no trailing whitespace",
    hash_algorithm: "SHA-256",
  };
}

export function canonicalizeFile(filepath, forceJson = null) {
  const raw = readFileSync(filepath);
  const pathStr = filepath.replace(/\\/g, "/");
  const record = hashCanonical(raw, pathStr, forceJson);
  record.canonical_bytes = canonicalize(raw, pathStr, forceJson);
  return record;
}

function selfTest() {
  const failures = [];

  const rawText = Buffer.from("hello  \r\nworld\t\r\n");
  const out = canonicalizeText(rawText);
  const expected = Buffer.from("hello\nworld\n");
  if (!out.equals(expected)) {
    failures.push(`text CRLF: got ${out.toString("hex")} expected ${expected.toString("hex")}`);
  }

  const bom = Buffer.from([0xef, 0xbb, 0xbf, ...Buffer.from("alpha\n")]);
  const outBom = canonicalizeText(bom);
  if (!outBom.equals(Buffer.from("alpha\n"))) {
    failures.push(`BOM: got ${outBom}`);
  }

  const j1 = Buffer.from('{"b": 2, "a": 1}\n');
  const j2 = Buffer.from('{\n  "a": 1,\n  "b": 2\n}\n');
  const c1 = canonicalizeJson(j1);
  const c2 = canonicalizeJson(j2);
  if (!c1.equals(c2)) {
    failures.push(`JSON order: ${c1} != ${c2}`);
  }
  if (!c1.equals(Buffer.from('{"a":1,"b":2}\n'))) {
    failures.push(`JSON compact: got ${c1}`);
  }

  const nested = Buffer.from('{"z": {"b": 1, "a": 2}, "y": [3, 1]}\n');
  const cn = canonicalizeJson(nested);
  if (!cn.equals(Buffer.from('{"y":[3,1],"z":{"a":2,"b":1}}\n'))) {
    failures.push(`nested JSON: got ${cn}`);
  }

  const d1 = sha256Hex(canonicalize(Buffer.from('{"x":1,"y":2}')));
  const d2 = sha256Hex(canonicalize(Buffer.from('{"y":2,"x":1}')));
  if (d1 !== d2) failures.push("determinism failed");

  const empty = canonicalizeText(Buffer.from(""));
  if (!empty.equals(Buffer.from(""))) failures.push(`empty: got ${empty}`);

  if (failures.length) {
    for (const f of failures) console.error("FAIL:", f);
    process.exit(1);
  }

  console.log("canonicalize self-test passed");
  const sample = canonicalize(Buffer.from('{"b":2,"a":1}'));
  console.log("sample_digest:", sha256Hex(sample));
  console.log("sample_bytes:", sample.toString());
}

const isMain = process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href;
if (isMain) {
  const args = process.argv.slice(2);
  if (args.length === 0) {
    selfTest();
  } else if (args[0] === "-h" || args[0] === "--help") {
    console.log("Usage:");
    console.log("  node canonicalize.mjs              # self-test");
    console.log("  node canonicalize.mjs <file> [...] # print path + sha256");
    console.log("  node canonicalize.mjs --json <file>");
    console.log("  node canonicalize.mjs --text <file>");
  } else {
    let force = null;
    let files = args;
    if (args[0] === "--json") {
      force = true;
      files = args.slice(1);
    } else if (args[0] === "--text") {
      force = false;
      files = args.slice(1);
    }
    for (const fp of files) {
      const rec = canonicalizeFile(fp, force);
      console.log(JSON.stringify({
        path: rec.path,
        sha256: rec.sha256,
        size_bytes: rec.size_bytes,
      }));
    }
  }
}

File: verification/canonicalize.py
#!/usr/bin/env python3
"""
CITIZEN_ROOT canonicalization.

Rules (from CITIZEN_ROOT_INDEX verification_rules):
  canonicalization: UTF-8, LF, sorted object keys, no trailing whitespace
  hash_algorithm:   SHA-256
  leaf_construction: SHA-256(canonical_bytes)
  rewrite_forbidden: true
  authority: false

Deterministic. No randomness. No authority.
"""

from __future__ import annotations

import hashlib
import json
import sys
from pathlib import Path
from typing import Any, Dict, Optional, Union


def sha256_hex(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


def _strip_trailing_ws_lines(text: str) -> str:
    """Normalize each line: strip trailing whitespace, join with LF."""
    lines = text.split("\n")
    stripped = [line.rstrip(" \t\r\f\v") for line in lines]
    body = "\n".join(stripped)
    if body and not body.endswith("\n"):
        body += "\n"
    if body.endswith("\n\n"):
        body = body.rstrip("\n") + "\n"
    return body


def canonicalize_text(raw: Union[str, bytes]) -> bytes:
    """
    Text path:
      - UTF-8 decode (strict)
      - normalize all line endings to LF
      - strip trailing whitespace on every line
      - ensure non-empty result ends with exactly one LF
      - no BOM
    """
    if isinstance(raw, bytes):
        if raw.startswith(b"\xef\xbb\xbf"):
            raw = raw[3:]
        text = raw.decode("utf-8")
    else:
        text = raw
        if text.startswith("\ufeff"):
            text = text[1:]

    text = text.replace("\r\n", "\n").replace("\r", "\n")
    text = _strip_trailing_ws_lines(text)
    return text.encode("utf-8")


def _sort_keys_recursive(obj: Any) -> Any:
    """Recursively sort object keys; leave arrays in order."""
    if isinstance(obj, dict):
        return {k: _sort_keys_recursive(obj[k]) for k in sorted(obj.keys())}
    if isinstance(obj, list):
        return [_sort_keys_recursive(v) for v in obj]
    return obj


def canonicalize_json(raw: Union[str, bytes, dict, list]) -> bytes:
    """
    JSON path:
      - parse (if needed)
      - recursively sort object keys
      - serialize compact: separators=(',', ':'), ensure_ascii=False
      - UTF-8, single trailing LF, no trailing whitespace
    """
    if isinstance(raw, (dict, list)):
        data = raw
    else:
        if isinstance(raw, bytes):
            if raw.startswith(b"\xef\xbb\xbf"):
                raw = raw[3:]
            text = raw.decode("utf-8")
        else:
            text = raw
            if text.startswith("\ufeff"):
                text = text[1:]
        text = text.replace("\r\n", "\n").replace("\r", "\n")
        data = json.loads(text)

    ordered = _sort_keys_recursive(data)
    serialized = json.dumps(
        ordered,
        ensure_ascii=False,
        sort_keys=True,
        separators=(",", ":"),
    )
    serialized = serialized.rstrip(" \t\r\n") + "\n"
    return serialized.encode("utf-8")


def detect_json(path: Optional[str], raw: bytes) -> bool:
    """Heuristic: .json extension or content starts with { or [ after whitespace."""
    if path and path.lower().endswith(".json"):
        return True
    sample = raw.lstrip()[:1]
    return sample in (b"{", b"[")


def canonicalize(
    raw: bytes,
    path: Optional[str] = None,
    force_json: Optional[bool] = None,
) -> bytes:
    """
    Apply the locked canonicalization rules.

    force_json:
      True  → always treat as JSON
      False → always treat as text
      None  → auto-detect
    """
    use_json = force_json if force_json is not None else detect_json(path, raw)
    if use_json:
        try:
            return canonicalize_json(raw)
        except (json.JSONDecodeError, UnicodeDecodeError):
            return canonicalize_text(raw)
    return canonicalize_text(raw)


def hash_canonical(
    raw: bytes,
    path: Optional[str] = None,
    force_json: Optional[bool] = None,
) -> Dict[str, Any]:
    """Return identity record for an artifact."""
    canonical = canonicalize(raw, path=path, force_json=force_json)
    digest = sha256_hex(canonical)
    return {
        "path": path,
        "sha256": digest,
        "size_bytes": len(canonical),
        "canonicalization": "UTF-8, LF, sorted object keys, no trailing whitespace",
        "hash_algorithm": "SHA-256",
    }


def canonicalize_file(filepath: str, force_json: Optional[bool] = None) -> Dict[str, Any]:
    """Read a file, canonicalize, return identity record + canonical bytes."""
    p = Path(filepath)
    raw = p.read_bytes()
    path_str = str(p).replace("\\", "/")
    record = hash_canonical(raw, path=path_str, force_json=force_json)
    record["canonical_bytes"] = canonicalize(raw, path=path_str, force_json=force_json)
    return record


def _self_test() -> None:
    failures = []

    raw_text = b"hello  \r\nworld\t\r\n"
    out = canonicalize_text(raw_text)
    expected = b"hello\nworld\n"
    if out != expected:
        failures.append(f"text CRLF: got {out!r} expected {expected!r}")

    bom = b"\xef\xbb\xbfalpha\n"
    out = canonicalize_text(bom)
    if out != b"alpha\n":
        failures.append(f"BOM: got {out!r}")

    j1 = b'{"b": 2, "a": 1}\n'
    j2 = b'{\n  "a": 1,\n  "b": 2\n}\n'
    c1 = canonicalize_json(j1)
    c2 = canonicalize_json(j2)
    if c1 != c2:
        failures.append(f"JSON order: {c1!r} != {c2!r}")
    if c1 != b'{"a":1,"b":2}\n':
        failures.append(f"JSON compact: got {c1!r}")

    nested = b'{"z": {"b": 1, "a": 2}, "y": [3, 1]}\n'
    cn = canonicalize_json(nested)
    if cn != b'{"y":[3,1],"z":{"a":2,"b":1}}\n':
        failures.append(f"nested JSON: got {cn!r}")

    d1 = sha256_hex(canonicalize(b'{"x":1,"y":2}'))
    d2 = sha256_hex(canonicalize(b'{"y":2,"x":1}'))
    if d1 != d2:
        failures.append("determinism failed")

    empty = canonicalize_text(b"")
    if empty != b"":
        failures.append(f"empty: got {empty!r}")

    if failures:
        for f in failures:
            print("FAIL:", f)
        sys.exit(1)

    print("canonicalize self-test passed")
    sample = canonicalize(b'{"b":2,"a":1}')
    print("sample_digest:", sha256_hex(sample))
    print("sample_bytes:", sample)


if __name__ == "__main__":
    if len(sys.argv) == 1:
        _self_test()
    elif sys.argv[1] in ("-h", "--help"):
        print("Usage:")
        print("  python3 canonicalize.py              # self-test")
        print("  python3 canonicalize.py <file> [...] # print path + sha256")
        print("  python3 canonicalize.py --json <file>")
        print("  python3 canonicalize.py --text <file>")
        sys.exit(0)
    else:
        force = None
        args = sys.argv[1:]
        if args[0] == "--json":
            force = True
            args = args[1:]
        elif args[0] == "--text":
            force = False
            args = args[1:]
        for fp in args:
            rec = canonicalize_file(fp, force_json=force)
            print(json.dumps({
                "path": rec["path"],
                "sha256": rec["sha256"],
                "size_bytes": rec["size_bytes"],
            }, sort_keys=True))
